### PR TITLE
WIP - DoAction - Load JSON file on demand and apply pre-processed patches 

### PR DESCRIPTION
**Pull request by duboisp**
Minor update

Change:
* Load JSON file on demand
* Pre-processed JSON patches based on variable from user input.

This change require to fully re-test every working example with the URL mapping.

It also fix an issue where the default value is not taken in consideration in some circumstance in the urlmapping working example. Now the behaviour match the description in the working example.

Todo:
* Review and ensure the English working example is complete
* Documentation of this new feature (2 new actions)
* Fully re-test the URL mapping
* Translate in French
* ESlinting
* ~~Fix the issue where Loading a JSON file and applying patches at the same time is buggy.~~
 (Fixed on 2019-09-09)

\cc @Paul-Doug

----
Edits....

* [ ] Add a console log error when withInput action is used without defining a "match" or a "default"

### CODE REVIEW

**Comment by duboisp** [comment]
A demo is available here: https://universallabs.org/wet/labs/pr-1594/unmin/demos/doaction/json-en.html
Download the corresponding build: https://universallabs.org/wet/labs/pr-1594.zip

**Comment by duboisp** [comment]
@Paul-Doug this new feature is now ready for the WET-BOEW contribution completion by your team. See the PR description for all the details and todo.

Live demo: https://universallabs.org/wet/labs/pr-1594/unmin/demos/doaction/json-en.html
GCWeb Build (same as this PR): https://universallabs.org/wet/labs/pr-1594.zip

Let me know if you have any question. I will be also available during the weekly Tuesday code sprint.

**Comment by Paul-Doug** [comment]
@duboisp I'm having some trouble when I replace a standard input field with a date type. The {base} and {qval} don't seem to be loaded. 

html:

<input type="date" id="mydatefield" min="2019-06-10" max="2019-06-12" value="">
</label>
<div>
<button type="button" data-wb-doaction='[
	{ "action": "loadJSON", "source": "#smallLoader", "url": "demo/prog-tracker-small.json" },
	
	{ "action": "withInput", "srcInput": "#mydatefield", "actions": [
		{
			"action": "patch", "source": "#smallLoader", "base": "/T1_digital",

			"patches" : {
				"op": "move",
				"path": "/date",
				"from": "{base}/{qval}"
			}
		}
	] }
]'>Apply a patch based on Date Input</button>
</div>

<div id="smallLoader" data-wb-jsonmanager='{
		"url": "demo/empty.json",
		"name": "jsonsmall",
		"wait": true,


		"debug": true
	}'></div>
	
<p>Data loaded or patched: <span data-json-replace="#[jsonsmall]/date">Date not loaded yet</span></p>


prog-tracker-small.json:

{
	"test" : "test found!",
	"T1_digital" : {
		"2019-06-10" : "2019-06-24", 
		"2019-06-11" : "2019-06-25", 
		"2019-06-12" : "2019-06-26"
	},
	"T1_paper" : {
		"2019-06-10" : "2019-06-26", 
		"2019-06-11" : "2019-06-27", 
		"2019-06-12" : "2019-06-28"
	}
}

**Comment by Paul-Doug** [comment]
@duboisp … code for my previous comment
[doaction-test.zip](https://github.com/wet-boew/GCWeb/files/3611326/doaction-test.zip)

**Comment by duboisp** [comment]
@Paul-Doug 
Your are missing to set a default value and to set a regular expression to parse the user input.

To solve your issue, you will need to update your JS file to include a default fall-back value, like:

```
{
	"test" : "test found!",
	"T1_digital" : {
		"bad-date": "X buisness days",
		"2019-06-10" : "2019-06-24", 
		"2019-06-11" : "2019-06-25", 
		"2019-06-12" : "2019-06-26"
	},
	"T1_paper" : {
		"bad-date": "X buisness days",
		"2019-06-10" : "2019-06-26", 
		"2019-06-11" : "2019-06-27", 
		"2019-06-12" : "2019-06-28"
	}
}
```

In your do action plugin configuration you will add the configuration named "match" and "default".
* "match" will contains a regular expression used to parse the user input. ex: ```^20[1,2][9,0]-[1,0][0-9]-[0,1,2,3][0-9]$```. I recommend to be specific
* "default" will contains a value in the case the user input do not match the regular expression.

I know I did it in my example, but defining the "base" configuration when we already specific the patch is not needed. We can directly put the value in the patch.

Example:
```
<button type="button" data-wb-doaction='[
	{ "action": "loadJSON", "source": "#smallLoader", "url": "demo/prog-tracker-small.json" },
	
	{ "action": "withInput", "srcInput": "#mydatefield", "actions": [
		{
			"action": "patch", "source": "#smallLoader", "match": "^20[1,2][9,0]-[1,0][0-9]-[0,1,2,3][0-9]$", "default": "bad-date",

			"patches" : {
				"op": "move",
				"path": "/date",
				"from": "/T1_digital/{qval}"
			}
		}
	] }
]'>Apply a patch based on Date Input</button>
```

Let me know if that work.

**Comment by Paul-Doug** [comment]
@duboisp 
Yes that works. Thank you!
So, is the "match" regex mandatory in this case? Or in all cases where "withInput" is used?

**Comment by duboisp** [comment]
> So, is the "match" regex mandatory in this case? Or in all cases where "withInput" is used?

Yes it is mandatory, If not specified, ```{[qval}``` and ```{base}``` keyword wont be replaced.

**Comment by Paul-Doug** [comment]
Is it possible to set the "base" parameter to a dynamic value? Maybe using a jQuery selector? or other?

For instance, for the JSON above could we have a field (id="myT1type") with a value of "T1_digital" or "T1_paper" and then set the patch something like:
_"action": "patch", "source": "#smallLoader", "base": "$('#myT1type')", ..._

This fails but I wonder if it's a syntax problem.

**Comment by Paul-Doug** [comment]
Or maybe URL Mapping is a way to do this?

**Comment by duboisp** [comment]
> Is it possible to set the "base" parameter to a dynamic value? 

No, only the qval is fully dynamic. But in your action, you can define a static value for "base". 

Making the "base" dynamic (or n variable dynamic) will require to add more configuration for the validation (matches and default). That will complexify how the do-action "withInput" can be configured.

**Comment by Paul-Doug** [comment]
To allow for different "from" values I'm trying to use an ajax action on a fieldflow. I think this should work but it fails to "default". @duboisp could you have a look?

[doaction-ajax.zip](https://github.com/wet-boew/GCWeb/files/3642753/doaction-ajax.zip)

**Comment by duboisp** [comment]
The feature described by this PR was documented and merged via PR #1602
